### DISCUSSION TITLE
Move from `PyMiniRacer` to `dukpy`?

### DISCUSSION THREAD

**Discussion by baseplate-admin** [emmyarty/dxsvelte#4] [Announcements]
Hey i have a small suggestion. Can we move from `PyMiniRacer` to `duktape`. Because it looks like `PyMiniRacer` is not updated in a long time. Whereas `dukpy` receives regular updates 

All in all thanks for this awesome library

**Comment by emmyarty** [comment]
It's true that it's been a while since PMR had an update, and I'm not super familiar with Dukpy / Duktape but it's still a bit  [patchy in fully implementing ES5/ES6+.](https://wiki.duktape.org/postes5features)

I'm open to moving to a more actively maintained V8 package, I just don't think Dukpy is the right fit since I'm not keen on adding legacy transpilers into the mix, especially since we're talking about the backend where we actually have some control over the runtime and environment.

PMR uses V8 (8.9) and supports ES12 in full. For the time being, I think this is the right choice. Thoughts?

**Comment by baseplate-admin** [comment]
> I just don't think Dukpy is the right fit since I'm not keen on adding legacy transpilers into the mix

Fair enough. I wont complain.

> I'm open to moving to a more actively maintained V8 package,

>  since we're talking about the backend where we actually have some control over the runtime and environment.

If you want to go this route may i suggest [this](https://github.com/denoland/rusty_v8) with [this](https://github.com/PyO3/pyo3/)?

> I think this is the right choice. 

Yep if we dont wanna support DukTape its better to use PyMiniRacer

**Comment by emmyarty** [comment]
The proposal sounds good. It won't be something I get around to in the near future since I know exactly zero Rust, but this might just be the excuse to start I've been looking for.

**Reply by baseplate-admin** [reply]
I encountered this today

https://github.com/Distributive-Network/PythonMonkey

Can be of value to this project

**Reply by emmyarty** [reply]
Sweet find! It looks like it's not quite at the MVP stage yet but for all intents it seems to work.

I also found this:
https://github.com/cloudflare/stpyv8

My inclination is to add support for both PythonMonkey and stPyV8, and then make it a simple setting to choose between the three options. Perhaps even include the ability to attempt the preferred one and then, if it fails, to attempt a backup engine. Failure throws an error, success on the reattempt throws a warning?

Ideally there would be just one, and it would remain bulletproof, but with all the upcoming changes happening to the Svelte compiler this could be the best way forward.

**Comment by andy0130tw** [comment]
I am exploring the possibility of bringing the SSR functionality of Svelte to other backend stack. I currently use a patched version of QuickJS and [a Python binding to it](https://github.com/PetterS/quickjs). It does the job like a charm! (Little runtime dependencies + ES2020 syntax support) If the regular maintenance is a concern, I know there is a JS runtime wrapper of it that is being actively developed, called [txiki.js](https://github.com/saghul/txiki.js). But I have not set it up for my project.

**Reply by baseplate-admin** [reply]
Sounds like a good plan. But supporting something mainstream ( v8, gecko ) is better for the maintainability and feature updates.

**Reply by emmyarty** [reply]
Hi, so I'm in agreement that the JS runtime will inevitably need to be changed, sadly it does look like QuickJS is in a similar boat to PyMiniRacer in that it hasn't been maintained for a couple of years now - in fact, the last release of QuickJS was a month before PMR's and it's behind one year more in its implementation of ES. I'm sure, like PMR, QuickJS is fantastic - but it's a little too like-for-like! :)
